Performance best practice update with davidfowl samples (#14734)
aspnet/AspNetCore.Docs: aspnetcore/performance/performance-best-practices.md

All IO in ASP.NET Core is asynchronous. Servers implement the `Stream` interface, which has both synchronous and asynchronous overloads. The asynchronous ones should be preferred to avoid blocking thread pool threads. Blocking threads can lead to thread pool starvation.

**Do not do this:** The following example uses the <xref:System.IO.StreamReader.ReadToEnd*>. It blocks the current thread to wait for the result. This is an example of [sync over async](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#warning-sync-over-async
).

[!code-csharp[](performance-best-practices/samples/3.0/Controllers/MyFirstController.cs?name=snippet1)]

In the preceding code, `Get` synchronously reads the entire HTTP request body into memory. If the client is slowly uploading, the app is doing sync over async. The app does sync over async because Kestrel does **NOT** support synchronous reads.

**Do this:** The following example uses <xref:System.IO.StreamReader.ReadToEndAsync*> and does not block the thread while reading.

[!code-csharp[](performance-best-practices/samples/3.0/Controllers/MyFirstController.cs?name=snippet2)]

> The preceding code asynchronously reads the entire HTTP request body into memory.

Reviewer: If we remove the ReadToEndAsync and show the async JSON reading directly then it's fine to remove this

* LOH is collected with the rest of the heap. LOH requires a full [garbage collection](/dotnet/standard/garbage-collection/fundamentals) or [Gen2 collection](/dotnet/standard/garbage-collection/fundamentals#generations).

This [blog post](https://adamsitnik.com/Array-Pool/#the-problem) describes the problem succinctly:

> When a large object is allocated, it’s marked as Gen 2 object. Not Gen 0 as for small objects. The consequences are that if you run out of memory in LOH, GC cleans up the whole managed heap, not only LOH. So it cleans up Gen 0, Gen 1 and Gen 2 including LOH. This is called full garbage collection and is the most time-consuming garbage collection. For many applications, it can be acceptable. But definitely not for high-performance web servers, where few big memory buffers are needed to handle an average web request (read from a socket, decompress, decode JSON & more).

Naively storing a large request or response body into a single `byte[]` or `string`:

* May result in quickly running out of space in the LOH.
* May cause performance issues for the app because of full GCs running.

> ## Use buffered and synchronous reads and writes as an alternative to asynchronous reading and writing

Reviewer 2: Working with a synchronous data processing API

> When using a serializer/de-serializer that only supports synchronous reads and writes:

Reviewer 2: like JSON.NET...

> * Choose to buffer the data into memory before passing data into the serializer/de-serializer.

Reviewer 2: ```suggestion
* Buffer the data into memory asynchronously before passing it into the serializer/de-serializer.
```

> * [JSON.NET](https://www.newtonsoft.com/json/help/html/Introduction.htm) is only synchronous.

Reviewer 2: This bullet point is in the wrong place. It's an example of a sync serializer, not an action to take when using one.

ASP.NET Core 3.0 uses <xref:System.Text.Json> by default for JSON serialization. <xref:System.Text.Json>:

* Reads and writes JSON asynchronously.
* Is optimized for UTF-8 text.
* Typically higher performance than `Newtonsoft.Json`.

## Do not store IHttpContextAccessor.HttpContext in a field

The [IHttpContextAccessor.HttpContext](xref:Microsoft.AspNetCore.Http.IHttpContextAccessor.HttpContext) returns the `HttpContext` of the active request when accessed from the request thread. The `IHttpContextAccessor.HttpContext` should **not** be stored in a field or variable.

**Do not do this:** The following example stores the `HttpContext` in a field, and then attempts to use it later.

[!code-csharp[](performance-best-practices/samples/3.0/MyType.cs?name=snippet1)]

The preceding code frequently captures a null or incorrect `HttpContext` in the constructor.

**Do this:** The following example:

* Stores the <xref:Microsoft.AspNetCore.Http.IHttpContextAccessor> in a field.
* Uses the `HttpContext` field at the correct time and checks for `null`.

[!code-csharp[](performance-best-practices/samples/3.0/MyType.cs?name=snippet2)]

> ## Do not access HttpContext from multiple threads

Reviewer: The highlights are messed up in this section.

![image](https://user-images.githubusercontent.com/95136/66107853-1e5fb080-e576-11e9-913d-ba58f6d7459e.png)

Reviewer: it's still busted https://review.docs.microsoft.com/en-us/aspnet/core/performance/performance-best-practices?view=aspnetcore-3.1&branch=pr-en-us-14734#do-not-access-httpcontext-from-multiple-threads

Author: It's correct on edge/FF and Chrome at 80%. The build team (who writes the HTML generator for all of doc.ms.com) broke it 6 months ago and have only managed to mostly fix it.

Edge:
![image](https://user-images.githubusercontent.com/3605364/66226387-62d27580-e676-11e9-92cf-496c45a33743.png)

FF:
![image](https://user-images.githubusercontent.com/3605364/66226410-71b92800-e676-11e9-922f-d870c6989f47.png)

Chrome at 80%:

![image](https://user-images.githubusercontent.com/3605364/66226460-8bf30600-e676-11e9-8909-de32958d8f43.png)

Author: @davidfowl  see #14936
I asked for the problem to be  prioritized on the internal bug.

[!code-csharp[](performance-best-practices/samples/3.0/Controllers/FireAndForgetSecondController.cs?name=snippet1)]

**Do this:** The following example:

* Injects an <xref:Microsoft.Extensions.DependencyInjection.IServiceScopeFactory> in order to create a scope in the background work item. `IServiceScopeFactory` is a singleton.
* Creates a new dependency injection scope in the background thread.
* Doesn't reference anything from the controller.
* Doesn't capture the `PokemonDbContext` from the incoming request.

[!code-csharp[](performance-best-practices/samples/3.0/Controllers/FireAndForgetSecondController.cs?name=snippet2)]

The following highlighted code:

* Creates a scope for the lifetime of the background operation and resolves services from it.
* Uses `PokemonDbContext` from the correct scope.

[!code-csharp[](performance-best-practices/samples/3.0/Controllers/FireAndForgetSecondController.cs?name=snippet2&highlight=9-16)]

> ## Avoid adding headers after the HttpResponse has started

Reviewer 2: ```suggestion
## Do not modify the status code or headers after the response body has started
```

Reviewer 2: Related: Do not call `next()` if you have already started writing to the response body. Components only expect to be called if it's possible for them to handle/manipulate the response.

ASP.NET Core does not buffer the HTTP response body. The first time the response is written:

* The headers are sent along with that chunk of the body to the client.
* It's no longer possible to change response headers.

**Do not do this:** The following code tries to add response headers after the response has already started:

[!code-csharp[](performance-best-practices/samples/3.0/Startup22.cs?name=snippet1)]

> In the preceding code, `context.Response.Headers["test"] = "test value";` if `next()` has written to the response.

Reviewer 2: ```suggestion
In the preceding code, `context.Response.Headers["test"] = "test value";` will throw an exception if `next()` has written to the response.
```

> **Do this:** The following example checks if the HTTP response has started before writing to the body.

Reviewer 2: ```suggestion
**Do this:** The following example checks if the HTTP response has started before modifying the headers.
```
